C++: New range analysis

rdmarsh2 opened the pull request:
This PR adds an IR-based port of the Java range analysis library, which has more precision than the existing SimpleRangeAnalysis  and supports bounds relative to other IR instructions and operands. The library will need more correctness tests as well as some performance testing before it's ready to merge.

The library currently only reports bounds relative to `Operand`s and `Instruction`s. A mapping layer to support the AST classes may be added in a followup PR.

A few features of the Java library are currently missing from the C++ reimplementation. The Java library supports range analysis across a small set of math functions. Those may be addressed as part of this PR or as a follow up PR. There is also a modulus analysis library used to improve precision in the Java range analysis, which will be ported in a separate PR.

kevinbackhouse commented on cpp/ql/src/semmle/code/cpp/rangeanalysis/RangeUtils.qll:
```diff
@@ -0,0 +1,54 @@
+predicate backEdge(PhiInstruction phi, PhiOperand op) {
+  phi.getAnOperand() = op and
+  (
+    phi.getBlock().dominates(op.getPredecessorBlock())
```
I don't think this is correct if the control flow graph isn't reducible. It works for Java, because Java doesn't have arbitrary gotos, so the CFG is always reducible in Java.

geoffw0 commented on cpp/ql/test/library-tests/rangeanalysis/rangeanalysis/RangeAnalysis.expected:
```diff
@@ -0,0 +1,2264 @@
+instructionBounds
```
These expected results are too long to review, and probably too long even to review a diff properly alongside future changes. Can we make a test that reports the ranges only of particular expressions of interest to a human - e.g. assignments to a variable with a particular name?

geoffw0 replied:
The new version of the test looks much better, thanks.

jbj commented on cpp/ql/test/library-tests/rangeanalysis/rangeanalysis/RangeAnalysis.ql:
```diff
@@ -0,0 +1,12 @@
+
+query predicate operandBounds(Operand op, Bound b, int delta, boolean upper, Reason reason) {
+  boundedOperand(op, b, delta, upper, reason)
+}
```
Why have separate test directories for `RangeAnalysis` and `SimpleRangeAnalysis`? If they share the same C/C++ test files, then it should be easier to compare the two analyses.

jbj commented on cpp/ql/src/semmle/code/cpp/rangeanalysis/RangeAnalysis.qll:
```diff
@@ -0,0 +1,590 @@
+ *
+ * An inferred bound can either be a specific integer, the abstract value of an
+ * SSA variable, or the abstract value of an interesting expression. The latter
+ * category includes array lengths that are not SSA variables.
```
This paragraph needs updating.

jbj commented on cpp/ql/src/semmle/code/cpp/rangeanalysis/RangeAnalysis.qll:
````diff
@@ -0,0 +1,590 @@
+ * ```
+ *   len = arr.length;
+ *   if (x < len) { ... y = x-1; ... y ... }
+ * ```
````
This comment block also needs updating. You can also mark it with a todo comment or delete it for now.

aschackmull replied:
Why does it need updating?  Because of the `arr.length` field access?  If that's a field in a struct then the example should still be fine and valid C code.

jbj replied:
Good point. It's fine as it is.

aschackmull commented on cpp/ql/src/semmle/code/cpp/rangeanalysis/RangeAnalysis.qll:
```diff
@@ -0,0 +1,590 @@
+import RangeAnalysisPublic
+
+/**
+ * Gets a condition that tests whether `i` equals `bound + delta` at `use`.
```
This comment doesn't match the predicate.  There's no `i` and no `use`.

aschackmull commented on cpp/ql/src/semmle/code/cpp/rangeanalysis/RangeAnalysis.qll:
```diff
@@ -0,0 +1,590 @@
+ * - `isEq = true`  : `i == bound + delta`
+ * - `isEq = false` : `i != bound + delta`
+ */
+private IRGuardCondition eqFlowCond(Operand op, Operand bound, int delta,
```
Hmm, this predicate doesn't seem to be referenced from anywhere useful? Also, I don't think I understand what this predicate does - at the very least it doesn't seem to match the corresponding java predicate in any way.

rdmarsh2 replied:
it looks like I didn't finish implementing `unequalOperand`, which would use it indirectly.

aschackmull commented on cpp/ql/src/semmle/code/cpp/rangeanalysis/RangeAnalysis.qll:
```diff
@@ -0,0 +1,590 @@
+  delta = 0
+  or
+  exists(IRGuardCondition guard |
+    guard = boundFlowCond(op2, op1, delta, upper, _) and
```
It looks like the conjunct `result.controls(op.getInstruction().getBlock(), testIsTrue)` has generally been folded into `boundFlowCond` when comparing to the java version.  Was there a particular reason for this?

rdmarsh2 replied:
That's because my `boundFlowCond` takes an `Operand` rather than an `Instruction`; I think it doesn't make sense to talk about an `Operand` being bounded by a guard that doesn't control it. It's possible that I should change `boundFlowCond` to take an `Instruction` and check the guard in `boundFlowStepSsa`.

aschackmull replied:
Right, but then the `boundFlowCond` predicate is actually part of the family of edge-defining predicates, so it should be named something including `step`, and the last parameter is no longer needed.  In fact it becomes more or less equal to `boundFlowStepSsa`, so it could be inlined.  Maybe collect these two predicates into one named something like `boundFlowStepCond` so it's clear that it's responsible for defining the edges that arise from conditions.

rdmarsh2 replied:
I'm going with the second suggestion; it simplifies eqFlowCond as well

aschackmull commented on cpp/ql/src/semmle/code/cpp/controlflow/IRGuards.qll:
````diff
@@ -268,6 +268,32 @@ class IRGuardCondition extends Instruction {
+     * return x;
+     * ```
+     */
+    cached predicate controlsEdge(ConditionalBranchInstruction branch, IRBlock succ, boolean testIsTrue) {
````
The "controls" part appears to be missing from the body of this predicate - as it's written it's only going to include inferences where the guard-successor leads directly to the phi-node.  Consider e.g. if the example above gets augmented with an unrelated block between the `x >= y` edge and the phi-node:
```
if (x < y) {
  x = y;
} else if (foo) {
  return 0;
}
return x;

aschackmull replied:
Found the missing part, as per comment below.  In that case, this predicate just has a misleading name.

rdmarsh2 replied:
renamed to `controlsEdgeDirectly`

aschackmull commented on cpp/ql/src/semmle/code/cpp/rangeanalysis/RangeAnalysis.qll:
```diff
@@ -0,0 +1,590 @@
+  or
+  exists(Operand compared |
+    result.comparesLt(compared, bound, delta, upper, testIsTrue) and
+    result.controls(op.getPredecessorBlock(), testIsTrue) and
```
Ah, I see now that this is where the "controls" that I mentioned was missing in `controlsEdge` ended up.

aschackmull commented on cpp/ql/src/semmle/code/cpp/rangeanalysis/RangeAnalysis.qll:
```diff
@@ -0,0 +1,590 @@
+ * range analysis.
+ */
+private predicate safeCast(IntegralType fromtyp, IntegralType totyp) {
+  fromtyp.getSize() <= totyp.getSize() and
```
Shouldn't this be strict in the case of casting from an unsigned type to a signed type?  E.g. I wouldn't imagine casting from `unsigned int` to `int` would qualify as a safe cast.

aschackmull commented on cpp/ql/src/semmle/code/cpp/rangeanalysis/RangeAnalysis.qll:
```diff
@@ -0,0 +1,590 @@
+    i1.(MulInstruction).hasOperands(op, c.getAUse()) and factor = k
+    or
+    exists(ShiftLeftInstruction i |
+      i = i1 and i.getAnOperand().(LeftOperand) = op and i.getAnOperand().(RightOperand) = c.getAUse() and factor = 2.pow(k)
```
It seems like maybe adding `getLeftOperand()` on `Instruction` as a short-hand for `.getAnOperand().(LeftOperand)` would be nice to have (resp. `getRightOperand()`).

aschackmull replied:
Or maybe I'm missing something here, because I immediately notice the use of `i.getRightOperand()` below.  Is there a difference between `i.getRightOperand()` and `i.getAnOperand().(RightOperand)`, seeing as the latter is used quite a bit?

rdmarsh2 replied:
Yes; originally, there wasn't an `Operand` IPA type, and so `getLeftOperand` and company get an `Instruction`. `getAnOperand().getLeftOperand()` gets the `Operand` object. I think we need to do some cleanup and renaming on that API, but probably not in this PR.

aschackmull commented on cpp/ql/src/semmle/code/cpp/rangeanalysis/RangeAnalysis.qll:
```diff
@@ -0,0 +1,590 @@
+    (if r1 instanceof NoReason then reason = r2 else reason = r1)
+  )
+  or
+  boundedInstruction(op.getDefinitionInstruction(), b, delta, upper, fromBackEdge, origdelta, reason)
```
This case here combined with the `op2.getDefinitionInstruction().getAnOperand().(CopySourceOperand) = op1` case in `boundFlowStepSsa` appears to duplicate some work.

rdmarsh2 replied:
Fixed by removing the case in boundFlowStepSsa

aschackmull commented on cpp/ql/src/semmle/code/cpp/rangeanalysis/RangeAnalysis.qll:
```diff
@@ -0,0 +1,590 @@
+  PhiOperand op, Bound b, int delta, boolean upper, boolean fromBackEdge, int origdelta,
+  Reason reason
+) {
+  exists(NonPhiOperand op2, int d1, int d2, Reason r1, Reason r2 | // should mid be an `Operand`?
```
> // should mid be an `Operand`?

That's a somewhat confusing comment.

rdmarsh2 replied:
I think I decided `mid` should be an `Operand`, changed the type, renamed it, and forgot to delete the comment

jbj commented on cpp/ql/src/semmle/code/cpp/rangeanalysis/Bound.qll:
```diff
@@ -0,0 +1,75 @@
+
+  predicate hasLocationInfo(string path, int sl, int sc, int el, int ec) {
+    path = "" and sl = 0 and sc = 0 and el = 0 and ec = 0
+  }
```
AFAICT the extractor always adds a `UnknownDefaultLocation`.

https://git.semmle.com/Semmle/code/blob/db7270cd01575786fe15bd63f482882a91207b3e/c/extractor/src/extractor_keys.cpp#L4111-L4112

rdmarsh2 replied:
Yeah, I just copied the JAva implementation; I've swapped to `UnknownDefaultLocation` here

aschackmull commented:
So, I've commented along the way as I read the code.  I think the code could be streamlined a bit and made easier to understand, since I felt like some of the predicates got slightly confusing in the translation from the java implementation.
The general algorithm works by propagating bound-facts along the edges of a graph, and the predicates are roughly divided into two groups - those that define the edges of the graph and those that propagate and update bound-facts (the latter group is easy to identify as they all have a paramenter of type `Bound`).  I think the former group of predicates (those defining the graph edges) can be made clearer.
To get a better grasp of this, we should consider exactly which nodes and which edges we need in the graph - and in particular avoid accidentally including edges that are unnecessary due to being the concatenation of two other edges.
In the java implementation there are 3 kinds of nodes: expressions, ssa variables, and ssa variable-ssa read position pairs. A read of an ssa variable (at some position) yield edges from ssa variables and ssa variable-read position pairs to expressions, most expressions yield expr-to-expr edges, ssa variable definitions yield edges from their defining expr to the variable, and guards comparing an ssa variable to some expr yield edges from that expr to all ssa variable-read position pairs that are corresponding ssa reads guarded by the condition.
In the C implementation there are as far as I've been able to gather just two kinds of nodes, `Instruction` and `Operand`. Edges go from operands to instructions, say in the case of an addition, or from an instruction to an operand (if the operand is defined by the instruction).  Then there's guards, and I'm not sure how to factor those edges in in the best way, but it should likely result in an operand-to-operand edge.

I hope this makes sense. :-)

rdmarsh2 commented on cpp/ql/src/semmle/code/cpp/rangeanalysis/Bound.qll:
```diff
@@ -0,0 +1,75 @@
+  TBoundOperand(Operand o) {
+    o.getDefinitionInstruction().getResultType() instanceof IntegralType or
+    o.getDefinitionInstruction().getResultType() instanceof PointerType
+  }
```
I had thought that bounding relative to `Operand` might give a tighter bound than `Instruction` in some cases involving guard conditions; I'm no longer convinced that that's true.

aschackmull replied:
It also looks like `TBoundInstruction` includes too much if it's trying to correspond to `TBoundSsa` - it should likely exclude temporary variables that are introduced by the IR construction.  Note, that the overall algorithm works by constructing a graph (that's all the `*step*` predicates), which should be fast as its size is likely linear in the size of the program, and the recursive calculation in the graph essentially calculates the transitive closure in the graph restricted to certain starting points, and the `Bound` class is exactly those starting points.  So `Bound` should be as small as possible to maintain good performance, while including all the interesting nodes in the graph that one might want as a bound for something else, and all phi nodes as they are needed as bounds for themselves to get any sort of relevant bounds in loops.

aschackmull replied:
I absolutely don't think value numbers in general should be part of `Bound` - that's going to be too big without any benefits.  E.g. bounding anything by the value number of `x+1` is never going to be useful, as it's always better to bound directly by `x`.

aschackmull replied:
Right, sorry, just now read the commit with the code change above to use value numbers.  When restricted to atomic things, then it looks like it might be fine.  I'm slightly concerned by the `i.getAUse() instanceof ArgumentOperand` disjunct - doesn't that open the door for valuenumbers of things like `x+1`?

rdmarsh2 replied:
Yes; Consider the following code:
``
int *foo(int x) {
  int *arr = new int[x*2];
  for(int i = 0; i < x*2; i++) {
    arr[i] = i; // we need to be able to relate i to the size of arr, which is x+1
  }
  return arr;
}
```
I think we will need to be able to have bounds relative to arbitrary `ValueNumber`s that are used as arguments to allocating functions, and given that C++ allows custom allocators, I don't think we'll be able to do that without allowing any `ValueNumber` used as an argument to be a bound.

aschackmull replied:
Ok, I can see why it can be useful.  Then let me just say that the size of the `Bound` type and whether a smaller set of valuenumbers can be used here is something to potentially revisit if performance becomes a problem.

rdmarsh2 replied:
I'm starting on performance testing; it looks like the main stage of range analysis is about 50% slower with the `i.getAUse() instanceof ArgumentOperand` disjunct than without.

geoffw0 commented on cpp/ql/test/library-tests/rangeanalysis/rangeanalysis/RangeAnalysis.expected:
```diff
@@ -0,0 +1,26 @@
+| test.cpp:37:6:37:10 | Load: begin | test.cpp:35:28:35:30 | InitializeParameter: end | 0 | true | CompareLT: ... < ... |
+| test.cpp:37:16:37:20 | Load: begin | test.cpp:35:28:35:30 | InitializeParameter: end | 0 | true | CompareLT: ... < ... |
+| test.cpp:38:5:38:11 | Load: ... ++ | test.cpp:35:28:35:30 | InitializeParameter: end | 0 | true | CompareLT: ... < ... |
+| test.cpp:44:13:44:13 | Load: y | test.cpp:42:29:42:29 | InitializeParameter: z | 0 | true | CompareLT: ... < ... |
```
Oh actually both bounds appear to be non-inclusive.  That's consistent.  No problem.

geoffw0 replied:
I think the rule is `lower bound <= value < upper bound`, can you confirm that?

aschackmull replied:
The library ought to compute bounds that are inclusive.  That goes for both upper and lower bounds.

geoffw0 replied:
Then the upper bound for the read of `y` on line 44 should be `z - 1`, right?

aschackmull replied:
Yes, I believe so.  There's a strengthening that's missing in `boundFlowCond` (I've also put a comment about it above in that predicate).

geoffw0 commented on cpp/ql/test/library-tests/rangeanalysis/rangeanalysis/RangeAnalysis.expected:
```diff
@@ -0,0 +1,26 @@
+| test.cpp:16:9:16:9 | Load: y | test.cpp:14:15:14:15 | InitializeParameter: x | 1 | false | CompareLT: ... < ... |
+| test.cpp:18:9:18:9 | Load: x | test.cpp:14:22:14:22 | InitializeParameter: y | 0 | false | CompareLT: ... < ... |
+| test.cpp:20:10:20:10 | Load: x | test.cpp:14:15:14:15 | InitializeParameter: x | -2 | false | NoReason |
+| test.cpp:20:10:20:10 | Load: x | test.cpp:14:22:14:22 | InitializeParameter: y | -2 | false | CompareLT: ... < ... |
```
Impressive results here, took me a while to figure out why they're correct. :+1:

aschackmull commented on cpp/ql/src/semmle/code/cpp/controlflow/IRGuards.qll:
````diff
@@ -268,6 +268,32 @@ class IRGuardCondition extends Instruction {
+     * return x;
+     * ```
+     */
+    predicate controlsEdgeDirectly(ConditionalBranchInstruction branch, IRBlock succ, boolean testIsTrue) {
````
I still think this is a misleading name.  I'd go with something like `hasBranchEdge`.  To me, "guard controls X" means that a certain branch edge from the guard dominates X in the sense that it must have been taken on all valid paths leading to X; this can then either manifest as "controls directly" when the dominance can be determined directly from the dominance graph, or more indirectly as in the following where `guard` being true controls X, without this being apparent from a naive dominance calculation, but needs a sequence of more complicated logical implications:
```
b = guard && foo;
if (b) { X }
```

rdmarsh2 replied:
Alright, changed in next push

aschackmull commented on cpp/ql/src/semmle/code/cpp/rangeanalysis/RangeAnalysis.qll:
```diff
@@ -0,0 +1,608 @@
+  boolean testIsTrue)
+{
+  exists(Operand compared |
+    result.comparesLt(compared, bound, delta, upper, testIsTrue) and
```
For integral types, this is missing the strengthening from `x < y` to `x <= y - 1`, right?  I noticed this because of the test result that made @geoffw0 ask about inclusive vs. non-inclusive bounds.

rdmarsh2 replied:
Fixed

aschackmull commented on cpp/ql/src/semmle/code/cpp/rangeanalysis/RangeAnalysis.qll:
```diff
@@ -124,14 +124,10 @@ import RangeAnalysisPublic
  * - `isEq = false` : `i != bound + delta`
  */
-private IRGuardCondition eqFlowCond(Operand op, Operand bound, int delta,
+private IRGuardCondition eqFlowCond(ValueNumber vn, Operand bound, int delta,
```
Much better!

aschackmull commented on cpp/ql/src/semmle/code/cpp/rangeanalysis/RangeAnalysis.qll:
```diff
@@ -160,37 +157,19 @@ private predicate boundFlowStepSsa(
+  exists(int d |
+    result.comparesLt(vn.getAUse(), bound, d, upper, testIsTrue) and
+    // strengthen from x < y to x <= y-1
+    if upper = true
```
This strengthening appears wrong.  There are at least four cases to be handled for each of the true-false combinations of `upper` and `testIsTrue`.  I'd think that in two of those cases we'd have `delta = d`, one case with `delta = d-1`, and one case with `delta = d+1`.

rdmarsh2 replied:
The C++ guards library's `comparelt` provides the same semantics as `boundFlowCond` but with a `<` rather than a `<=` when `upper = true`. I've added a test to demonstrate.

aschackmull replied:
Ah, right.

aschackmull commented on cpp/ql/src/semmle/code/cpp/rangeanalysis/RangeAnalysis.qll:
```diff
@@ -564,6 +549,7 @@ private predicate boundedInstruction(
     boundedPhiCandValidForEdge(i, b, delta, upper, fromBackEdge, origdelta, reason, op)
   )
   or
+  not i instanceof PhiInstruction and
```
Uhm, is this right? We need the base case of a phi variable being bounded by itself.

rdmarsh2 replied:
Fixed

aschackmull commented on cpp/ql/src/semmle/code/cpp/rangeanalysis/RangeAnalysis.qll:
```diff
@@ -0,0 +1,594 @@
+import RangeAnalysisPublic
+
+/**
+ * Gets a condition that tests whether `op` equals `bound + delta`.
```
The qldoc should refer to `vn` instead of `op` and `i`.

aschackmull commented on cpp/ql/src/semmle/code/cpp/rangeanalysis/RangeAnalysis.qll:
```diff
@@ -0,0 +1,594 @@
+}
+
+/**
+ * Gets a condition that tests whether `op` is bounded by `bound + delta`.
```
s/op/vn/

rdmarsh2 commented:
I've retargeted to `next` for the moment because I need #648 for termination. Initial performance testing looks acceptable but not great; 4 minutes of evaluation after building IR on ChakraCore. Half of that time is before the main stage of the range analysis, which probably means there's performance improvements to be made in the sign analysis library.  That time was with a 1.19 snapshot, my changes on top of `next`, and a 1.18 evaluator, so the performance is likely better with a 1.19 evaluator.

I've done a bit of glancing at stats on ChakraCore; about half of instructions have a bound, about a tenth of those have a non-self bound. There's 370 instructions after pruning by #648 that have an impossible pair of bounds (upper and lower bounds relative to the same value number, with the lower bound's delta > the upper bound's delta). I suspect that most of those are actually unreachable, but I haven't done a manual review yet.

aschackmull commented on cpp/ql/src/semmle/code/cpp/rangeanalysis/RangeAnalysis.qll:
```diff
@@ -0,0 +1,606 @@
+ *
+ * If the condition evaluates to `testIsTrue`:
+ * - `isEq = true`  : `i == bound + delta`
+ * - `isEq = false` : `i != bound + delta`
```
s/i/vn/

aschackmull commented on cpp/ql/src/semmle/code/cpp/rangeanalysis/RangeAnalysis.qll:
```diff
@@ -0,0 +1,606 @@
+  )
+}
+
+/** Holds if `v != e + delta` at `pos`. */
```
qldoc needs an update.

aschackmull commented on cpp/ql/src/semmle/code/cpp/rangeanalysis/RangeAnalysis.qll:
```diff
@@ -0,0 +1,606 @@
+
+/** Holds if `v != e + delta` at `pos`. */
+private predicate unequalFlowStep(
+ Operand op1, Operand op2, int delta, Reason reason
```
I'd suggest switching the names of `op1` and `op2` - that would make it consistent with all the other predicates by having the flow of inference go from `op1` to `op2`.

aschackmull commented on cpp/ql/src/semmle/code/cpp/rangeanalysis/RangeAnalysis.qll:
```diff
@@ -0,0 +1,606 @@
+  exists(Operand op2, int d1, int d2 |
+    unequalFlowStep(op, op2, delta, reason) and
+    boundedNonPhiOperand(op2, b, d2, true, _, _, _) and
+    boundedNonPhiOperand(op2, b, d2, true, _, _, _) and
```
This is quite wrong. There are two identical conjuncts (one should have s/true/false/); `d1` is not bound (the argument to `unequalFlowStep` should be `d1` instead of `delta`).

aschackmull commented on cpp/ql/src/semmle/code/cpp/rangeanalysis/RangeAnalysis.qll:
```diff
@@ -0,0 +1,606 @@
+private predicate boundedInstruction(
+  Instruction i, Bound b, int delta, boolean upper, boolean fromBackEdge, int origdelta, Reason reason
+) {
+  isReducibleCFG(i.getFunction()) and
```
Presumably this is temporary until back edges are identified during their construction?

aschackmull commented:
A few comments to your observations:
The sign analysis in java was quite prone to bad join-orders, so you may experience something similar.
For unreachable code, it is expected that the range analysis can come up with impossible pairs of bounds (as that contradiction might be the reason the code is unreachable).  As an example, in the following code, it is reasonable to infer both `x < y` and `x > y` in the body of the `if`:
```
if (x < y && x > y) { ... }
```

rdmarsh2 commented:
> The sign analysis in java was quite prone to bad join-orders, so you may experience something similar.

That seems likely to be the problem in sign analysis

>For unreachable code, it is expected that the range analysis can come up with impossible pairs of bounds 

Right; I'm just thinking that that's likely to be fertile ground for finding correctness issues.

aschackmull commented on cpp/ql/src/semmle/code/cpp/rangeanalysis/RangeAnalysis.qll:
```diff
@@ -161,7 +161,11 @@ private IRGuardCondition boundFlowCond(ValueNumber vn, NonPhiOperand bound, int
   exists(int d |
     result.comparesLt(vn.getAUse(), bound, d, upper, testIsTrue) and
-    // strengthen from x < y to x <= y-1
+    /*
```
A minor style comment: I'd go with `//`-style comments for in-predicate comments, and leave multiline `/* .. */` for top-level things.  (This is also what the autoformatter currently expects, so this would add a spurious newline when autoformatted)

jbj replied:
Agreed. In-predicate comments should be easy to ignore and should therefore not stand out visually.

jbj commented:
To what extent is it possible to compare this new range analysis to SimpleRangeAnalysis to look for improvements and regressions? Can you make a test that uses the sample code from both the old and the new analyses and produces output with four columns for each `Expr`: its {upper, lower} bound with the {old, new} range analysis. Only rows where the analyses give different results need to be included.

rdmarsh2 commented:
It looks like most of the differences result from type-derived bounds. Most of these are cases where the old library gets bounds of `MAXINT-1` or `MININT+1` because it ignores overflow while doing arithmetic with unbounded `int`s. However, some other differences are regressions where a type-derived bound is relevant after widening; line 371 in test.c is an example of this.

There are a few cases where a constant bound isn't propagated across a negation by the new library. We can't do anything useful with negation for relative bounds, but I could add something to handle it for constant bounds.

rdmarsh2 commented:
[Comparison.txt](https://github.com/Semmle/ql/files/2728778/Comparison.txt)

rdmarsh2 commented:
For the initial off-by-one query we're planning to write using this library, I think we should be OK with the current results. If we're going to try and replace `SimpleRangeAnalysis.qll`, we'll need to handle the type-derived bounds (hopefully in a more accurate way) as well as negations. I also noticed one more type of regression - we don't give a lower bound for remainders where both operands are positive.

rdmarsh2 commented:
... and that's just an instance of a broader problem with replacing `SimpleRangeAnalysis`: the algorithm we're using doesn't combine constant bounds from both operands of an arithmetic operation, so we won't get a constant upper bound on `x + y` in the following case:
```
if (x < 10) {
  if (y < 10) {
    return x + y;
  }
}
```

rdmarsh2 commented:
@jbj  and I discussed the possible constant bounds improvements on a call this morning; we would like to do those in the future and deprecate `SimpleRangeAnalysis.qll`, but we decided not to do them as part of this PR. I think I've addressed all feedback, and I'm removing the WIP label; Jonas and Schack should feel free to merge at this point.

jbj commented:
The tests failed. I'm guessing the failures come from the switch from 1.18 to 1.19, so please check them carefully as they may be symptoms of problems with 1.19.

rdmarsh2 commented:
Sign analysis and range analysis changes look like improvements - I think they're related to better CFG pruning after the rebase. GVN changes also look like improvements, but I'm not sure what the change was. Alias analysis improvements, maybe?

rdmarsh2 commented:
@jbj the failing test is qlcfg, but I didn't expect anything I've touched to affect it... Can you take a look?

jbj commented:
That's a semantic merge conflict that's crept in because the submodule pointer is too old to include this test. I've opened #747 to fix it.
